File: apps/androidtv/core/designsystem/src/main/kotlin/dev/spidola/tv/core/designsystem/PosterRail.kt
// SPDX-FileCopyrightText: 2026 Spidola contributors
// SPDX-License-Identifier: AGPL-3.0-or-later

@file:Suppress("MatchingDeclarationName") // Named for the public composable; the data type is a helper.

package dev.spidola.tv.core.designsystem

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyRow
import androidx.compose.foundation.lazy.items
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.focus.focusRestorer
import androidx.compose.ui.unit.dp
import androidx.tv.material3.ClickableSurfaceDefaults
import androidx.tv.material3.MaterialTheme
import androidx.tv.material3.Surface
import androidx.tv.material3.Text
import kotlinx.collections.immutable.ImmutableList

/**
 * One card in a [PosterRail]. [id] is a caller-composed stable key (e.g. source+identity), so it
 * doubles as the lazy-list key that `focusRestorer` needs and the focus key across a data refresh.
 */
data class PosterItem(
    val id: String,
    val title: String,
    val subtitle: String?,
    val logo: String?,
)

/**
 * A titled, horizontally-scrolling rail of poster cards — the home screen's favorites and recents
 * rows (PRD §8.3). D-pad focus moves card to card on the foundation [LazyRow] with pivot scrolling;
 * `focusRestorer` with stable keys restores focus on back-navigation. An empty rail renders nothing
 * so the caller can omit the section entirely.
 */
@Composable
fun PosterRail(
    title: String,
    items: ImmutableList<PosterItem>,
    onSelect: (PosterItem) -> Unit,
    modifier: Modifier = Modifier,
) {
    if (items.isEmpty()) return
    Column(
        modifier = modifier,
        verticalArrangement = Arrangement.spacedBy(SpidolaSpacing.s),
    ) {
        Text(
            text = title,
            style = MaterialTheme.typography.titleLarge,
            color = SpidolaPalette.BroadcastWhite,
            modifier = Modifier.padding(horizontal = SpidolaSpacing.safeHorizontal),
        )
        LazyRow(
            modifier = Modifier.fillMaxWidth().focusRestorer(),
            contentPadding = PaddingValues(horizontal = SpidolaSpacing.safeHorizontal),
            horizontalArrangement = Arrangement.spacedBy(SpidolaSpacing.m),
        ) {
            items(items = items, key = { it.id }) { item ->
                PosterCard(item = item, onClick = { onSelect(item) })
            }
        }
    }
}

@Composable
private fun PosterCard(
    item: PosterItem,
    onClick: () -> Unit,
) {
    Surface(
        onClick = onClick,
        modifier = Modifier.width(CARD_WIDTH.dp),
        shape = ClickableSurfaceDefaults.shape(shape = SpidolaFocus.cardShape),
        colors =
            ClickableSurfaceDefaults.colors(
                containerColor = SpidolaPalette.Set,
                contentColor = SpidolaPalette.BroadcastWhite,
            ),
        scale = SpidolaFocus.scale(),
        border = SpidolaFocus.border(),
    ) {
        Column {
            LogoImage(
                url = item.logo,
                modifier = Modifier.fillMaxWidth().aspectRatio(CARD_ASPECT),
            )
            Text(
                text = item.title,
                style = MaterialTheme.typography.labelMedium,
                color = SpidolaPalette.BroadcastWhite,
                maxLines = 1,
                modifier = Modifier.padding(SpidolaSpacing.s),
            )
        }
    }
}

private const val CARD_WIDTH = 240
private const val CARD_ASPECT = 16f / 9f

File: apps/tvos/Packages/DesignSystem/Sources/DesignSystem/PosterRail.swift
// SPDX-FileCopyrightText: 2026 Spidola contributors
// SPDX-License-Identifier: AGPL-3.0-or-later

import SwiftUI

/// One card in a `PosterRail`. `id` is a caller-composed stable key (e.g. source+identity) so it
/// doubles as the `ForEach` identity and the focus key across a data refresh.
public struct PosterItem: Identifiable, Hashable, Sendable {
  public let id: String
  public let title: String
  public let subtitle: String?
  public let logo: String?

  public init(id: String, title: String, subtitle: String? = nil, logo: String? = nil) {
    self.id = id
    self.title = title
    self.subtitle = subtitle
    self.logo = logo
  }
}

/// A titled, horizontally-scrolling rail of poster cards — the home screen's favorites and recents
/// rows (PRD §8.3). D-pad focus moves card to card, each wearing the Test-Card Amber treatment;
/// the native focus engine handles pivot scrolling. Empty rails render nothing so the caller can
/// omit the section entirely.
public struct PosterRail: View {
  private let title: String
  private let items: [PosterItem]
  private let onSelect: (PosterItem) -> Void

  @FocusState private var focused: String?

  public init(title: String, items: [PosterItem], onSelect: @escaping (PosterItem) -> Void) {
    self.title = title
    self.items = items
    self.onSelect = onSelect
  }

  public var body: some View {
    if items.isEmpty {
      EmptyView()
    } else {
      VStack(alignment: .leading, spacing: SpidolaSpacing.s) {
        Text(title)
          .font(SpidolaType.title)
          .foregroundStyle(SpidolaPalette.broadcastWhite)
          .padding(.horizontal, SpidolaSpacing.safeHorizontal)
        ScrollView(.horizontal, showsIndicators: false) {
          LazyHStack(spacing: SpidolaSpacing.m) {
            ForEach(items) { item in
              PosterCard(item: item, isFocused: focused == item.id) { onSelect(item) }
                .focused($focused, equals: item.id)
            }
          }
          .padding(.horizontal, SpidolaSpacing.safeHorizontal)
          .padding(.vertical, SpidolaSpacing.s)
        }
      }
    }
  }
}

private struct PosterCard: View {
  let item: PosterItem
  let isFocused: Bool
  let action: () -> Void

  private let cardWidth: CGFloat = 240

  var body: some View {
    Button(action: action) {
      VStack(alignment: .leading, spacing: SpidolaSpacing.xs) {
        LogoImage(url: item.logo)
          .frame(width: cardWidth, height: cardWidth * 9 / 16)
          .background(SpidolaPalette.set)
        Text(item.title)
          .font(SpidolaType.caption)
          .foregroundStyle(SpidolaPalette.broadcastWhite)
          .lineLimit(1)
        if let subtitle = item.subtitle {
          Text(subtitle)
            .font(SpidolaType.caption)
            .foregroundStyle(SpidolaPalette.staticGray)
            .lineLimit(1)
        }
      }
      .frame(width: cardWidth, alignment: .leading)
    }
    .buttonStyle(.plain)
    .spidolaFocusRing(isFocused: isFocused)
    .accessibilityLabel(
      item.subtitle.map { String(localized: "\(item.title), \($0)", bundle: .module) }
        ?? item.title)
  }
}
